### PR TITLE
feat: add rma shipping options

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- Adds support for RMA shipping options on swaps

### Why
- We want to allow customer service to choose between and customize shipping options when creating a swap

### How
**Models**
- adds an `RMAShippingOption` model and repository
- updates `Swap` and `ClaimOrder` models to include an `rma_shipping_options` one-to-many relationship

**Services**
- modifies `SwapService.create` method to include a `rmaShippingOptions` argument used in the creation of a swap
- adds a `CartService.addRMAMethod` responsible for adding the correct shipping method (either from a normal or rma option) to a cart
- tweaks `CartService.addShippingMethod` to include a custom price. This is used in case we need to add a method from a `RMAShippingOption` 
- adds a `ShippingProfileService.fetchRMAOptions`

**API**
- updates `POST /store/carts/:id/shipping-methods/` to use the cart service's `addRMAMethod` instead of `addShippingMethod`. In case the option does not correspond to an RMAShippingOption, then `addRMAMethod` will simply call `addShippingMethod`
- updates `POST /admin/orders/:id/swap` to accept a `rma_shipping_options` list request payload
- updates `GET /store/shipping-options/:cart_id/` to either call `shippingProfileService.fetchRMAOptions` in case `cart.type == swap || claim` or `shippingProfileService.fetchCartOptions` in case the cart is not a swap nor a claim

### Testing
- integration test to ensure the creation of a swap with `rma_shipping_options`
- integration test to ensure the addition of a normal and rma shipping method to a cart
- integration test to ensure the retrieval of shipping options for both a swap and default cart
- unit tests for API changes and new service methods

### CODE REVIEW

File: packages/medusa/src/api/routes/store/shipping-options/list-shipping-options.js [hunk 1]
```diff
@@ -40,7 +40,12 @@ export default async (req, res) => {
       relations: ["region", "items", "items.variant", "items.variant.product"],
     })
 
-    const options = await shippingProfileService.fetchCartOptions(cart)
+    let options
+    if (cart.type === "swap" || cart.type === "claim") {
+      options = await shippingProfileService.fetchRMAOptions(cart)
+    } else {
+      options = await shippingProfileService.fetchCartOptions(cart)
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/shipping-options/list-shipping-options.js:47]: Similar to how we approach adding a shipping method, would it be an idea to move this to the service layer?

File: packages/medusa/src/migrations/1632851018347-add_rma_shipping_options.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        await queryRunner.query(`CREATE INDEX "IDX_f3618cd0930a2e7357eddbebf4" ON "rma_shipping_option" ("shipping_option_id") `);
+        await queryRunner.query(`CREATE INDEX "IDX_4e4c2c0a3223c79a84a6f54e58" ON "rma_shipping_option" ("swap_id") `);
+        await queryRunner.query(`CREATE INDEX "IDX_2e262bb0e324b501d80a69f9df" ON "rma_shipping_option" ("claim_order_id") `);
+        await queryRunner.query(`ALTER TYPE "public"."cart_type_enum" RENAME TO "cart_type_enum_old"`);
+        await queryRunner.query(`CREATE TYPE "public"."cart_type_enum" AS ENUM('default', 'swap', 'draft_order', 'payment_link', 'claim')`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" DROP DEFAULT`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" TYPE "public"."cart_type_enum" USING "type"::"text"::"public"."cart_type_enum"`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" SET DEFAULT 'default'`);
+        await queryRunner.query(`DROP TYPE "public"."cart_type_enum_old"`);
```
Comment (olivermrbl) [inline packages/medusa/src/migrations/1632851018347-add_rma_shipping_options.ts:16]: Remove `public.`

File: packages/medusa/src/migrations/1632851018347-add_rma_shipping_options.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    public async down(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "rma_shipping_option" DROP CONSTRAINT "FK_2e262bb0e324b501d80a69f9df4"`);
+        await queryRunner.query(`ALTER TABLE "rma_shipping_option" DROP CONSTRAINT "FK_4e4c2c0a3223c79a84a6f54e583"`);
+        await queryRunner.query(`ALTER TABLE "rma_shipping_option" DROP CONSTRAINT "FK_f3618cd0930a2e7357eddbebf4c"`);
+        await queryRunner.query(`CREATE TYPE "public"."cart_type_enum_old" AS ENUM('default', 'swap', 'draft_order', 'payment_link')`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" DROP DEFAULT`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" TYPE "public"."cart_type_enum_old" USING "type"::"text"::"public"."cart_type_enum_old"`);
+        await queryRunner.query(`ALTER TABLE "public"."cart" ALTER COLUMN "type" SET DEFAULT 'default'`);
+        await queryRunner.query(`DROP TYPE "public"."cart_type_enum"`);
+        await queryRunner.query(`ALTER TYPE "public"."cart_type_enum_old" RENAME TO "cart_type_enum"`);
```
Comment (olivermrbl) [inline packages/medusa/src/migrations/1632851018347-add_rma_shipping_options.ts:31]: Remove `public.`

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Great! 

But thinking, we might have gone a step too far, too early with the RMA shipping options for claims. Currently, we don't have a clear idea of, how we want to extend the claims functionality with the RMA options (i take the blame for including it in the description in the first place). Therefore, we should perhaps restrict the RMA options to only work for swaps for now. Would love your input on this @srindom.

**Review by srindom — CHANGES_REQUESTED** [review]
A thought that comes to mind looking over this PR and which also addresses the question raised by @olivermrbl:

Should we modify the RMA relations to only belong to Carts i.e. when creating a Swap the resulting RMA options would have:

```
RMA option
- option_id
- price
- cart_id <-- corresponds to swap.cart_id
```

Doing it this way would mean that we wouldn't have to replicate functionality across swaps + claims + potential future entities, as long as these entities can use the Cart API for processing. It would also have the added benefit of giving store operators the option of specifying custom shipping options in relation to Draft Orders.

Given that we agree that this is a good idea we should probably reconsider the naming as RMA options would likely be a too narrow scope (it already is if we allow these to be applied to draft order carts). Maybe something like `Custom Shipping Options` would make sense, but happy to hear suggestions.

Please let me know what you think? :)

File: packages/medusa/src/services/cart.js [hunk 4]
```diff
@@ -1368,6 +1373,35 @@ class CartService extends BaseService {
...
+      if (rmaOption) {
+        await this.addShippingMethod(cartId, rmaOption.shipping_option_id, {
+          ...data,
+          price: rmaOption.price,
+        })
+      } else {
+        await this.addShippingMethod(cartId, optionIdOrRmaOptionId, data)
+      }
+    })
+  }
```
[the remark this answers is not in the document]
Comment (zakariaelas) [reply packages/medusa/src/services/cart.js:1403]: love the suggestion! so true!

File: packages/medusa/src/services/cart.js [hunk 5]
```diff
@@ -1368,6 +1373,35 @@ class CartService extends BaseService {
...
+   * the cart.
+   * @param {Object} cart - the cart object
+   * @param {string} optionIdOrCustomOptionId - id of the normal or custom shipping option to add as valid method
+   * @returns {{ optionId: string; customPrice: { price: number; } | {};}}
+   */
+  extractShippingOptionIdAndPrice(cart, optionIdOrCustomOptionId) {
+    if (
+      cart.custom_shipping_options &&
+      cart.custom_shipping_options.length > 0
+    ) {
```
Comment (olivermrbl) [inline packages/medusa/src/services/cart.js:1387]: Would prefer `if(cart.custom_shipping_options?.length)`

File: packages/medusa/src/services/shipping-profile.js [hunk 6]
```diff
@@ -413,6 +413,11 @@ class ShippingProfileService extends BaseService {
    * @return {[ShippingOptions]} a list of the available shipping options
    */
   async fetchCartOptions(cart) {
+    const customShippingOptions = cart.custom_shipping_options
+
+    if (customShippingOptions && customShippingOptions.length > 0)
```
Comment (olivermrbl) [inline packages/medusa/src/services/shipping-profile.js:418]: Same as above: `customShippingOptions?.length`

Comment (olivermrbl) [reply packages/medusa/src/services/shipping-profile.js:418]: And would love brackets :)

Comment (olivermrbl) [reply packages/medusa/src/services/shipping-profile.js:418]: We could do
```
if (cart.custom_shipping_options?.length) {
    return cart.custom_shipping_options
}
```

**Comment by zakariaelas** [comment]
@srindom I feel like we might run into some issues given how we implement the “Give free shipping” option. To re-iterate, selecting ‘give free option’ will also set the digital copy and printed copy as shipping options. If there is no way to filter out these, I think it’s good to tell your clients so that it makes sense :slightly_smiling_face:
image

**Comment by srindom** [comment]
@zakariaelas would it be possible to get filter them when calling `fetchCartOptions` - something like:
```
const profileIds = productsInCart.map(({ profile_id }) => profile_id)
const optionsToReturn = allOptions.filter(o => profileIds.includes(o.profile_id ))
`

**Pull request merged**